engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceWithResultTest.java

```diff
@@ -160,7 +160,8 @@ public void shouldSendRejectionImmediately() {
   }
 
   private void interceptResponseWriter(final CommandResponseWriter mockCommandResponseWriter) {
-    doAnswer(
+    when(mockCommandResponseWriter.valueWriter(any()))
```

Comment:
:thinking: Interesting. There is a significant change now here, since we call the `mockCommandResponseWriter.valueWriter` directly. 

Normally this happens because of concurrent access between stubbing and calling the spy. I guess this happens because of the @BeforeClass and @Before. In the BeforeClass we trigger something which will later call the responseWriter, where we in the @Before reset and add a stub. 

Was it a specific test which failed or the complete class?

Reply:
`shouldSendRejectionImmediately` repeatedly fails:

```
08:46:08.052 [Broker-0-StreamProcessor-1] ERROR io.camunda.zeebe.processor - Expected to execute side effects for record 'LoggedEvent [type=0, version=0, streamId=1, position=1, key=-1, timestamp=1665125168027, sourceEventPosition=-1] RecordMetadata{recordType=COMMAND, intentValue=255, intent=CREATE, requestStreamId=-2147483648, requestId=-1, protocolVersion=3, valueType=DEPLOYMENT, rejectionType=NULL_VAL, rejectionReason=, brokerVersion=8.2.0}' successfully, but exception was thrown.
java.lang.NullPointerException: Cannot invoke "io.camunda.zeebe.engine.api.CommandResponseWriter.partitionId(int)" because the return value of "io.camunda.zeebe.engine.api.CommandResponseWriter.rejectionType(io.camunda.zeebe.protocol.record.RejectionType)" is null
	at io.camunda.zeebe.streamprocessor.ProcessingStateMachine.lambda$executeSideEffects$13(ProcessingStateMachine.java:423) ~[classes/:?]
	at io.camunda.zeebe.scheduler.retry.ActorRetryMechanism.run(ActorRetryMechanism.java:28) ~[classes/:?]
	at io.camunda.zeebe.scheduler.retry.AbortableRetryStrategy.run(AbortableRetryStrategy.java:45) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:92) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:45) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:119) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:106) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:87) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:198) ~[classes/:?]
08:46:08.053 [] INFO  io.camunda.zeebe.test - Test failed, following records were exported:

org.mockito.exceptions.misusing.UnfinishedStubbingException: 
Unfinished stubbing detected here:
-> at io.camunda.zeebe.engine.processing.processinstance.CreateProcessInstanceWithResultTest.interceptResponseWriter(CreateProcessInstanceWithResultTest.java:163)

E.g. thenReturn() may be missing.
Examples of correct stubbing:
    when(mock.isOk()).thenReturn(true);
    when(mock.isOk()).thenThrow(exception);
    doThrow(exception).when(mock).someVoidMethod();
Hints:
 1. missing thenReturn()
 2. you are trying to stub a final method, which is not supported
 3. you are stubbing the behaviour of another mock inside before 'thenReturn' instruction is completed


	at io.camunda.zeebe.engine.processing.processinstance.CreateProcessInstanceWithResultTest.interceptResponseWriter(CreateProcessInstanceWithResultTest.java:176)
	at io.camunda.zeebe.engine.processing.processinstance.CreateProcessInstanceWithResultTest.setUp(CreateProcessInstanceWithResultTest.java:56)
	at jdk.internal.reflect.GeneratedMethodAccessor28.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.RunBefores.invokeMethod(RunBefores.java:33)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:30)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
```

Reply:
Ok give me 5 minutes I try something.

Reply:
I think I get it now. The engine is shared between test methods but the `@Before` sets up mocking for every test method. I guess if the engine from a previous test method is still trying to do stuff when `interceptResponseWriter` is called for the next test, Mockito is throwing this failure.

Thanks for double checking, I was also wondering why switching from `doAnswer/when` to `when/thenAnswer` would work. (Actually I'm still not sure why it would "fix" or hide the issue).

I'll try and make some changes so that `interceptResponseWriter` does not interfere with previous tests.

Reply:
@oleschoenburg exactly my thought. I just moved 
```

  @BeforeClass
  public static void init() {
    mockCommandResponseWriter = ENGINE.getCommandResponseWriter();
    interceptResponseWriter(mockCommandResponseWriter);
    ENGINE.deployment().withXmlResource(PROCESS).deploy();
  }
```

And this also works. You need to make then some things static of course.